src/operation/insert.rs

```diff
@@ -54,46 +60,26 @@ impl<'a> OperationWithDefaults for Insert<'a> {
     const NAME: &'static str = "insert";
 
     fn build(&mut self, description: &StreamDescription) -> Result<Command<Self::Command>> {
+        let max_doc_size: usize = Checked::new(description.max_bson_object_size).try_into()?;
+        let max_message_size: usize =
+            Checked::new(description.max_message_size_bytes).try_into()?;
+        let max_operations: usize = Checked::new(description.max_write_batch_size).try_into()?;
+
+        let mut command_body = rawdoc! { Self::NAME: self.ns.coll.clone() };
+        let options = bson::to_raw_document_buf(&self.options)?;
+        extend_raw_document_buf(&mut command_body, options)?;
+
+        let max_document_sequence_size =
+            max_message_size - OP_MSG_OVERHEAD_BYTES - command_body.as_bytes().len();
+
         let mut docs = Vec::new();
-        let mut size = 0;
-
-        let max_doc_size = Checked::<usize>::try_from(description.max_bson_object_size)?;
-        let max_doc_sequence_size =
-            Checked::<usize>::try_from(description.max_message_size_bytes)? - COMMAND_OVERHEAD_SIZE;
-
-        for (i, &d) in self
-            .documents
-            .iter()
-            .take(Checked::new(description.max_write_batch_size).try_into()?)
-            .enumerate()
-        {
-            let mut doc = d.to_owned();
-            let id = match doc.get("_id")? {
-                Some(b) => b.try_into()?,
-                None => {
-                    let mut new_doc = RawDocumentBuf::new();
-                    let oid = ObjectId::new();
-                    new_doc.append("_id", oid);
-
-                    let mut new_bytes = new_doc.into_bytes();
-                    new_bytes.pop(); // remove trailing null byte
-
-                    let mut bytes = doc.into_bytes();
-                    let oid_slice = &new_bytes[4..];
-                    // insert oid at beginning of document
-                    bytes.splice(4..4, oid_slice.iter().cloned());
-
-                    // overwrite old length
-                    let new_length = Checked::new(bytes.len()).try_into::<i32>()?.to_le_bytes();
-                    bytes[0..4].copy_from_slice(&new_length);
-                    doc = RawDocumentBuf::from_bytes(bytes)?;
-
-                    Bson::ObjectId(oid)
-                }
-            };
+        let mut current_size = 0;
```

Comment:
Suggest using `Checked` to prevent possible unsigned overflow when adding to `current_size`:

```suggestion
        let mut current_size = Checked::new(0);
```